java/ql/lib/ext/threat-grouping.model.yml

```diff
@@ -0,0 +1,56 @@
+extensions:
+
+  # An incomplete set of sources database reads
+  - addsTo:
+      pack: codeql/java-all
+      extensible: sinkModel
+    data:
+      - ["java.io", "OutputStream", True, "write", "(byte[])", "", "Argument[0]", "sql", "manual"]
+      - ["java.io", "OutputStream", True, "write", "(byte[],int,int)", "", "Argument[0]", "sql", "manual"]
+      - ["java.io", "OutputStream", True, "write", "(int)", "", "Argument[0]", "sql", "manual"]
+
+
+  - addsTo:
+      pack: codeql/java-all
+      extensible: sourceModel
+    data:
+      # Create a few
+      - ["java.sql", "PreparedStatement", True, "executeQuery", "()", "", "ReturnValue", "sql", "manual"]
+      - ["java.sql", "PreparedStatement", True, "getMetaData", "()", "", "ReturnValue", "sql", "manual"]
+      - ["java.sql", "PreparedStatement", True, "getParameterMetaData", "", "", "ReturnValue", "sql", "manual"]
+      - ["java.sql", "Statement", True, "executeQuery", "(String)", "", "ReturnValue", "sql", "manual"]
+      - ["java.sql", "Statement", True, "getResultSet", "()", "", "ReturnValue", "sql", "manual"]
+      - ["java.sql", "Statement", True, "getGeneratedKeys", "()", "", "ReturnValue", "sql", "manual"]
+      - ["java.sql", "Statement", True, "getConnection", "()", "", "ReturnValue", "sql", "manual"]
+      - ["java.sql", "ResultSet", True, "getInt", "(int)", "", "Argument[0]", "sql", "manual"]
+
+      # This one is not defined elsewhere. Why not? I can't get my example working without it.
+      - ["java.io", "InputStream", True, "read", "()", "", "ReturnValue", "remote", "manual"]
```

Comment:
Is this source correct?

Reply:
As a *real* source I would say no. This would mean that *any* local implementation of `InputStream` would also behave as a source - that is probably to general.